enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,138 @@
+---
+
+# cluster-runtime-constraint
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+This proposal presents the specification and usage to enable runtime constraints to be considered during operator dependency resolution. This feature will ensure Operator Lifecycle Manage (OLM) to install operators that satisfy dependency requirements and runtime constraints if present.
+
+## Motivation
+
+At the moment, OLM resolves a set of operators which seem that they will work together based on dependency requirements that are specifed by operator bundles. However, it is possible that those resolved operators will not work on the clusters due to cluster/runtime constraints. For example, if the cluster is running a specific Kubernetes version that is not compatible with the operators, the installed operators may fail to work propertly on that given cluster.
+
+The cluster runtime constraints are often unknown by the operator authors and should be specified by cluster admins. However, currently, there is no machanism for cluster admins to specify runtime constraints that OLM can understand and take under consideration during installation.
+
+### Goals
+- Provide machanism (including specification and guideline) for cluster admins to specify cluster runtime constraints
+- Provide specification and guideline for author operators to specify runtime contraints/requirements if needed
+- Block the updates/installations of operators based on cluster runtime constraints
+- Allow the cluster runtime constraints to be generic and confugurable by cluster admins
+
+### Non-Goals
+
+- Provide machanism to retrieve/populate cluster runtime constraints automatically
+- Allow cluster runtime constraints to be scoped
+- Allow cluster runtime constraints to come from multiple/different sources
+- Allow optional cluster runetime constraints
+
+
+## Proposal
+
+To provide a generic solution for cluster runtime constaints, a library named [cel-go](https://github.com/google/cel-go) that provides expression evaluation capabilities is utilized in this proposal.
+
+### User Stories
+
+#### Constraints specified by cluster admin
+
+As a cluster admin, I would like to specify a list of cluster runtime constraints that are associated with a given cluster and all installable operators will meet these requirements.
+
+#### Constraint properties specified by operator author
+
+As an operator author, I would like to provide a list of runtime properties that may be needed to be considered before installation to ensure the operator working properly on the cluster.
+
+### Implementation Details
+
+#### Cluster Runtime Constraint Configmap
+
+Cluster admins can specified cluster runtime constraints in a configmap (named `olm-runtime-constraints`) in `olm` namespace using this format:
```

Comment:
I did envision this to be a new API (CRD) that has configmap as the backend storage or in the CRD's spec itself. I would expect that part will be implemented later along with the auto detection. It shouldn't be a problem for OLM to support this singleton configmap for now and will add support for a new API later. In the end of the day, the data is what matters the most to OLM and I expect OLM to support multiple sources of cluster constraints eventually.

Reply:
The hard part is the migration. OLM will presumably have to support both APIs (configmap and your future CRD) and/or provide a migration path from one to the other. If this proposal includes an assumption that the API will fundamentally change, it would be a good idea to state that explicitly and add details of how that migration will be handled.

Reply:
I would also argue that using a configmap does not exempt us from supporting and maintaining compatibility with whatever goes in it as an API.

Reply:
I agree with @mhrivnak here. What we're really saying here is this:

> We're going to use a ConfigMap for our runtime cluster constraints API, and do runtime validation of its contents.

I think there's actually a lot _more_ complexity for us AND for users by putting this in a ConfigMap.

With a CRD:
- users get the benefit of self-documenting schema definitions and `kubectl explain` to help them understand how to specify constraints. They also get immediate feedback from the API server if they have an invalid object.
- OLM can encode its validation logic in CRD schema and/or a validating webhook, which means less runtime checking and error handling. For example, what would we do when some constraints are valid and other constraints aren't? Fail the entire ConfigMap or just the invalid constriants. How would we communicate to a cluster admin which constraints are invalid?

Reply:
+1 on using CRD